### PR TITLE
add close connection to avoid thousands of conns open

### PR DESCRIPTION
**Pull request by maxadamo**
I have noticed too many connections open and I found that the ping test was not closing the connections.

### CODE REVIEW

**Comment by maxadamo** [comment]
resolves #31

**Comment by toni-moreno** [comment]
Hello @maxadamo  sorry for the delay in this answer, and thank you very much for your contribution.

Prior to merge I would like to discuss a little about this "close()". I did no review all the code, and right now I have not enough time to spend in   testing this changes.  I will be grateful if you can do these test yourself.

I will add comments on the code.

File: pkg/agent/influxmonitor.go
Comment (toni-moreno) [inline pkg/agent/influxmonitor.go:78]: Here We are assigning the connection handler if everything ok. this sounds like we are reusing this connection , on other place in the code. ( sorry I can remember right now exactly where, Could you locate where we are using `im.lastcli`and why ?)  If reusing connection you can not close here... if not reusing connection perhaps could be nice  also removing `im.lastcli= con` assignments. Isn't it?

**Comment by maxadamo** [comment]
Before making this change, a new connection was added every 10 seconds (the check interval) and it was never released. I found 7 thousand connections open. 
Now it works, and I see only one connection. 
**p.s.**: when the the recover happens a new connection is added and it's not being closed... we may need to find where the other `Close()` is missing, but this is less dangerous, because the recover does not happen every 10 seconds and we can take it easier.

**Review by toni-moreno — CHANGES_REQUESTED** [review]
Hi @maxadamo  , I agree we should fix this connection leak issue.

But I'm reviewing code  and have doubts this can be work on sync/replication time.

Sync actions are done with the returned cli , in the InitPing() that you are forcing close,

https://github.com/toni-moreno/syncflux/blob/344a32867c21b511ef4563208c7da0cb0e5ab637/pkg/agent/influxmonitor.go#L90-L92

As you can see here on  initCluster() method Master and Slave DB's are assigned with the returned cli

https://github.com/toni-moreno/syncflux/blob/4ddf6a212a6735a31352c29b6d8024b06100d574/pkg/agent/agent.go#L93-L98

And on each recovery/sync detection the cli is updated. with the last cli with the UpdateCli() method and Sync data with ReplicateData()

https://github.com/toni-moreno/syncflux/blob/344a32867c21b511ef4563208c7da0cb0e5ab637/pkg/agent/hacluster.go#L254-L260

Where the update cli happens here.

https://github.com/toni-moreno/syncflux/blob/344a32867c21b511ef4563208c7da0cb0e5ab637/pkg/agent/influxmonitor.go#L107-L112

The ReplicateData()  here does not make any OpenCli() it reuses from the previously assigned

https://github.com/toni-moreno/syncflux/blob/344a32867c21b511ef4563208c7da0cb0e5ab637/pkg/agent/hacluster.go#L158-L170

I can not understand how can connect to the slave  DB (after it has been restarted) if the previously taken cli on the InitPing, has been closed, perhaps should be better close connection ( if exist) before reopen it , isn't it ? , and leave opened current connection for later use ( until recheck again, where will be closed again previously reopened. Did you understand me? Could you test closing before?. 

I'm really surprised with this Close() you can recover / sync data , ( maybe , sorry I wrote code a lot of time and should review again twice before any merge).

File: pkg/agent/influxmonitor.go
Comment (maxadamo) [reply pkg/agent/influxmonitor.go:78]: hi @toni-moreno 
I understand what you're saying. What I did, should have gone wrong, but `im.lastcli` is used only within this file and the recovery really worked: I had 3 lines less, and the gap was filled. 
I'll need to dig a bit more :face_with_head_bandage:

Comment (maxadamo) [reply pkg/agent/influxmonitor.go:78]: the functions, returns the followings: 
```golang
return con, dur, ver, nil
```

`con` is used inside the file agent.go, here:
```golang
cli, _, _, err := MDB.InitPing()
```  

And this is going back in the file influxmonitor.go to run this function (which should create a lock): 
```golang
func (im *InfluxMonitor) SetCli(cli client.Client) {
	im.climutex.Lock()
	defer im.climutex.Unlock()
	im.cli = cli
}
```

perhaps we need to add something like this, and initiate a new connection: 
```golang
con, err2 := client.NewHTTPClient(info)
```

but I'm still a bit confused.

Comment (toni-moreno) [reply pkg/agent/influxmonitor.go:78]: The cleanest way to close leaked connections i doing it just before reassing in `SetCli()` and `UpdateCli()`

```golang
If im.cli != nil {
  Im.cli.Close()
}
im.cli = cli
// or
im.cli = lastcli
```
Could you test it please?

Comment (maxadamo) [reply pkg/agent/influxmonitor.go:78]: Not working. After few minutes I am already getting over 200 connections:
```bash
lsof -nPi | grep syncflux | wc -l
237
```

**Comment by toni-moreno** [comment]
Hi @maxadamo still thinking on this issue. 
In a few weeks I will do some tests and if no better solution found I will merge your PR.

**Pull request merged**
